### PR TITLE
default a CSR's allowed usage to key encipherment and digital signing

### PR DESCRIPTION
**Pull request by mikedanese**
Some pretty safe and sane defaults.

@liggitt

### CODE REVIEW

File: pkg/apis/certificates/v1alpha1/defaults.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import "k8s.io/kubernetes/pkg/runtime"
+
+func addDefaultingFuncs(scheme *runtime.Scheme) error {
+	RegisterDefaults(scheme)
+	return scheme.AddDefaultingFuncs(
+		SetDefaults_CertificateSigningRequestSpec,
+	)
+}
+func SetDefaults_CertificateSigningRequestSpec(obj *CertificateSigningRequestSpec) {
+	if len(obj.Usages) == 0 {
```
Comment (liggitt) [inline pkg/apis/certificates/v1alpha1/defaults.go]: Maybe only if nil? An explicitly specified empty list seems like an invalid CSR

File: pkg/apis/certificates/v1alpha1/defaults.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (liggitt) [inline pkg/apis/certificates/v1alpha1/defaults.go]: 2017?

**Comment by liggitt** [comment]
Did kubelet CSR usage get updated to request a client cert usage yet?

File: pkg/apis/certificates/v1alpha1/defaults.go (hunk 1 above)
Comment (mikedanese) [inline pkg/apis/certificates/v1alpha1/defaults.go]: Done.

Comment (liggitt) [inline pkg/apis/certificates/v1alpha1/defaults.go]: thanks, validation check for nil usages would probably be helpful

Comment (mikedanese) [inline pkg/apis/certificates/v1alpha1/defaults.go]: Done.

File: pkg/apis/certificates/validation/validation.go [hunk 3]
```diff
@@ -50,8 +50,14 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
...
 	allErrs := apivalidation.ValidateObjectMeta(&csr.ObjectMeta, isNamespaced, ValidateCertificateRequestName, field.NewPath("metadata"))
 	err := validateCSR(csr)
+
+	specPath := field.NewPath("spec")
+
 	if err != nil {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("request"), csr.Spec.Request, fmt.Sprintf("%v", err)))
+		allErrs = append(allErrs, field.Invalid(specPath.Child("request"), csr.Spec.Request, fmt.Sprintf("%v", err)))
+	}
+	if csr.Spec.Usages == nil {
```
Comment (liggitt) [inline pkg/apis/certificates/validation/validation.go]: with defaulting, should never be nil, and clients that don't specify should never be empty. I was thinking a zero-length check here should make it invalid

Comment (mikedanese) [inline pkg/apis/certificates/validation/validation.go]: Why wouldn't we just default the zero length array then if it is invalid?

Comment (liggitt) [inline pkg/apis/certificates/validation/validation.go]: old clients that don't know about the field send nil, so we default them.

new clients that know about the field and explicitly specify an empty list should get an error

Comment (mikedanese) [inline pkg/apis/certificates/validation/validation.go]: I understand now. Fixed.

**Comment by liggitt** [comment]
LGTM

**Pull request merged**